File: ABC311/C.py
N = int(input())
A = [i - 1 for i in list(map(int, input().split()))]
visited = [False] * N
prev = [-1] * N

is_all_cycle = False
is_cycle = [False] * N
ans = []


def dfs(v, idx):
    global is_all_cycle
    if visited[v] == idx:
        is_cycle[v] = True
        is_all_cycle = True
        return

    if visited[v]:
        return

    visited[v] = idx
    prev[A[v]] = v
    dfs(A[v], idx)

    if not is_all_cycle:
        return

    if is_cycle[v]:
        now = v
        while True:
            ans.append(now)
            now = prev[now]

            if now == v:
                break

        print(len(ans))
        print(*[i + 1 for i in ans[::-1]])
        exit()


for i in range(N):
    if visited[i]:
        continue

    dfs(i, i+1)

File: ABC311/C.cpp
/* Code By [ junah ]
GitHub : [ junah201 ] */
#include <bits/stdc++.h>
using namespace std;
#define all(v) v.begin(), v.end()
#define int long long int
#define INF LONG_LONG_MAX

int N;
vector<int> graph;
vector<char> visited;

stack<int> ans;
bool is_cycle = false;
bool end_cycle = false;
int cycle_start = -1;

void dfs(int now, int idx)
{
    visited[now] = idx;

    int next = graph[now];

    if (visited[next] == idx)
    {
        ans.emplace(now);
        is_cycle = true;
        cycle_start = next;
        return;
    }

    if (visited[next] == 0)
    {
        dfs(next, idx);
    }

    if (is_cycle && !end_cycle)
    {
        ans.emplace(now);
        if (now == cycle_start)
        {
            end_cycle = true;
        }
    }
}

signed main()
{
    scanf("%lld", &N);
    graph.resize(N + 1, -1);
    visited.resize(N + 1, false);
    for (int i = 1; i <= N; i++)
    {
        scanf("%lld", &graph[i]);
    }

    for (int i = 1; i <= N; i++)
    {
        if (!visited[i])
        {
            dfs(i, i);
            if (!ans.empty())
            {
                printf("%lld\n", ans.size());
                while (!ans.empty())
                {
                    printf("%lld ", ans.top());
                    ans.pop();
                }
                return 0;
            }
        }
    }

    return 0;
}